src/cli/expand-patterns.js

```diff
@@ -0,0 +1,221 @@
+      } else if (stat.isDirectory()) {
+        entries.push({
+          type: "dir",
+          path: absolutePath,
+          glob: getSupportedFilesGlob(),
+          input: pattern
+        });
+      }
+    } else if (pattern[0] === "!") {
+      // convert negative patterns to `ignore` entries
+      globbyOptions.ignore.push(pattern.slice(1));
+    } else {
+      entries.push({
+        type: "glob",
+        // Using backslashes in globs is probably not okay, but not accepting
+        // backslashes as path separators on Windows is even more not okay.
+        // https://github.com/prettier/prettier/pull/6776#discussion_r380723717
+        // https://github.com/mrmlnc/fast-glob#how-to-write-patterns-on-windows
+        glob: isWindows ? pattern.replace(/\\/g, "/") : pattern,
+        input: pattern
+      });
+    }
+  }
+
+  for (const { type, path, glob, input } of entries) {
+    switch (type) {
+      case "file":
+        yield path;
+        continue;
+
+      case "dir": {
+        let result;
+        try {
+          result = globby.sync(glob, { ...globbyOptions, cwd: path });
+        } catch ({ message }) {
+          yield { error: `Unable to expand directory: ${input}\n${message}` };
+          continue;
+        }
+
+        if (result.length === 0) {
+          yield {
+            error: `No supported files were found in the directory "${input}".`
+          };
+        } else {
+          yield* sortIfNotBig(result);
+        }
+        continue;
+      }
+
+      case "glob": {
+        let result;
+        try {
+          result = globby.sync(glob, globbyOptions);
+        } catch ({ message }) {
+          yield {
+            error: `Unable to expand glob pattern: ${input}\n${message}`
+          };
+          continue;
+        }
+
+        if (result.length === 0) {
+          yield {
+            error: `No files matching the pattern "${input}" were found.`
+          };
+        } else {
+          yield* sortIfNotBig(result);
+        }
+        continue;
+      }
+    }
+  }
+
+  function getSupportedFilesGlob() {
+    if (!supportedFilesGlob) {
+      const extensions = flat(
+        context.languages.map(lang => lang.extensions || [])
+      );
+      const filenames = flat(
+        context.languages.map(lang => lang.filenames || [])
+      );
+      supportedFilesGlob = `**/{${extensions
+        .map(ext => "*" + (ext[0] === "." ? ext : "." + ext))
+        .concat(filenames)}}`;
+    }
+    return supportedFilesGlob;
+  }
+}
+
+/**
+ * @param {string} absolutePath
+ * @param {string} cwd
+ * @param {Record<string, boolean>} ignoredDirectories
+ */
+function containsIgnoredPathSegment(absolutePath, cwd, ignoredDirectories) {
+  return path
+    .relative(cwd, absolutePath)
+    .split(path.sep)
+    .some(dir => ignoredDirectories[dir]);
+}
+
+/**
+ * @param {string[]} paths
+ */
+function sortIfNotBig(paths) {
+  return paths.length > 10000
```

Comment:
Can you clarify magic number?

Reply:
I'm not sure about this actually. It should probably be much bigger, like 1e7. What if Prettier is used to format millions of files? Wouldn't this sorting cause a useless slowdown?

Reply:
Thinking more about it, now I see that it's a premature optimization. I'll remove this condition. Let's fix it when we face it.

Reply:
I think we always need to sort here, because we have to be deterministic, yes, it can be slow, for formatting really very very very big projects we should implement multithreading (maybe plus caching) (developers can write a custom script already).

Reply:
BTW, the sync API of `fast-glob` [returns results in a deterministic order](https://github.com/sindresorhus/globby/issues/131#issuecomment-590094878).